docs/sdg/sdg-refactor.md

```diff
@@ -0,0 +1,136 @@
+
+### Option 2: Move preprocessing and postprocessing into a new repository
+
+We could have a new repository for preprocessing and postprocessing and move all the preprocessing and postprocessing code there.
+
+Pro:
+
+- Avoids all the cons of Option 1.
+- Preprocessing and postprocessing are a coherent pieces of functionality that *could* have their own library (or libraries, FWIW).
+
+Cons:
+
+- Avoids all the pros of Option 1.
+- Having a separate repository with its own library brings in an enormous amount of overhead in maintaining that repository (e.g., CI/CD).
+- Having a separate repository with its own library also brings in an enormous amount of overhead in maintaining the CLI repository's dependency on all of those libraries.
+- Does not allow user flow 2.1 but maybe that's OK because it is not a priority and anyway the users could approximate that flow by also installing the ingestion library.
+
+Conclusion:
+
+- The cost of having a separate reposity is so high that we would only consider this option as a last resort.
+
+### Option 3: Move preprocessing and postprocessing into the CLI repository
+
+Pro:
+
+- The CLI already has a lot of "supporting" (non-core) functionality, so it would respect established precedent to include preprocessing and/or postprocessing here.
+- Supporting user flow 1.3.2 requires separate CLI commands for preprocessing and core SDG.  This is slightly simpler if preprocessing is implemented in CLI.  If preprocessing remains in the SDG library instead then the CLI would need to make _separate_ calls to the SDG library for preprocessing and core SDG to support user flow 1.3.2.  That adds a little complexity. 
+- Avoids some of the cons of Option 1, but see below for some overlap.
+- Avoids some of the cons of Option 2, but see below for some overlap.
+
+Con:
+
+- Avoids the pros of both Option 1 and Option 2.
+- As with Option 1, this approach involves a lot of coordination.  There are a lot of stakehoders involved in the CLI and locating preprocessing and postprocessing there drags those stakeholders into issues relating to preprocessing and postprocessing.  However, as with Option 1, coordinating across stakeholders is something an open source project needs to do well anyway to remain engaged with the community. 
+- As with Option 1, this approach suffers from the fact that keeping interconnected components in the same repository provides less pressure to consistently document the API contracts between them.  In the case of Option 1, the interconnected components that would not have as much pressure to be documented would be preprocessing/postprocessing and core SDG.  In the case of Option 3, the interconnected components that would not have as much pressure to be documented would be the CLI and preprocessing/postprocessing.  However, in both cases, this con could be alleviated by just having the discipline to document the APIs well even without such pressure.
```

Comment:
I believe that this option forces more documentation and contract maintenance than option 1.

Having the CLI own pre and post processing will enforce the SDG loop to remain compatible with the user contracts set by the `prep`, `generate`, and `process` commands. If all code lives in `instructlab-sdg` we risk siloing and only focusing on the "modular" usecases that team commonly deals with for the product.

Reply:
Right, but having the CLI own pre and post processing doesn't force documentation and contract maintenance between the CLI itself and the functions that implement these features in the CLI.  That's easily mitigated by doing the documentation and contract maintenance between the CLI and functions even without being forced to do so.  But it is a point of concern that I think mirrors the equivalent point of concern if it is all kept in SDG.